js/keyword-search.js

```diff
@@ -0,0 +1,212 @@
+namespacer('baltimoreCounty');
+
+baltimoreCounty.keywordSearch = (function keywordSearch($, sessionStorage, Handlebars) {
+	'use strict';
+
+	var searchData;
+
+	var documentClickHandler = function documentClickHandler() {
+		var $searchResults = $('#header-search-results');
+
+		if ($searchResults.is(':visible')) {
+			$searchResults.hide();
+		}
+	};
+
+	/**
+	 * Highlights the matched term in the results.
+	 * 
+	 * @param {String} searchTerm 
+	 * @param {Array<Term, Order>} matches 
+	 */
+	var highlightMatches = function highlightMatches(searchTerm, matches) {
+		var highlightedMatches = [];
+
+		matches.forEach(function forEachTopFiveMatch(match) {
+			var highlightedMatch = $.extend({}, match);
+			highlightedMatch.Term = highlightedMatch.Term.replace(searchTerm, '<strong>' + searchTerm + '</strong>');
+			highlightedMatches.push(highlightedMatch);
+		});
+
+		return highlightedMatches;
+	};
+
+	var init = function init(callback) {
+		if (sessionStorage && sessionStorage.searchData) {
+			searchData = JSON.parse(sessionStorage.searchData);
+		} else {
+			$.ajax('/sebin/m/m/searchTerms.json')
+				.then(onDataLoadedHandler, onDataLoadedError);
+		}
+
+		if (typeof callback !== 'undefined') {
+			callback();
+		}
+	};
+
+
+	/**
+	 * Handles the loaded data and puts it in session storage in the browser.
+	 * 
+	 * @param {Array<Term, Order>} data 
+	 */
+	var onDataLoadedHandler = function onDataLoadedHandler(data) {
+		searchData = data;
+		sessionStorage.setItem('searchData', JSON.stringify(data));
+	};
+
+	/**
+	 * Error handler
+	 * 
+	 * @param {Object} err 
+	 */
+	var onDataLoadedError = function onDataLoadedError(err) {
+		console.log(err);
+	};
+
+	/**
+	 * Matches the almighty Google's autocomplete rules
+	 * 
+	 * @param {string} searchTerm 
+	 * @param {Array<Term, Order>} matches 
+	 */
+	var orderByNameThenPopularity = function orderByNameThenPopularity(searchTerm, matches) {
+		if (matches.length === 0 || matches.length === 1) {
+			return matches;
+		}
+
+		var orderedMatches = [];
+
+		matches.forEach(function eachMatch(match) {
+			if (match.Term.indexOf(searchTerm) === 0) {
```

Comment:
are we concerned that ```match``` or ```match.Term``` might not exist?

Reply:
Added a check for Object.prototype.hasOwnProperty.call(match, 'Term')